src/main/haskell/kore/src/Kore/SMT/SMT.hs

```diff
@@ -0,0 +1,171 @@
+) where
+
+import           Data.Default
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+
+import Kore.AST.Common
+import Kore.AST.MetaOrObject
+import Kore.AST.PureML
+import Kore.ASTUtils.SmartPatterns
+import Kore.ASTUtils.Substitution
+import Kore.Attribute.Parser
+       ( ParseAttributes (..) )
+import Kore.Builtin.Hook
+import Kore.IndexedModule.MetadataTools
+
+import Data.Reflection
+import Data.SBV
+
+import GHC.IO.Unsafe
+
+data SMTAttributes
+  = SMTAttributes
+  { hook :: !Hook
+  } deriving(Eq, Show)
+
+instance Default SMTAttributes where
+    def = SMTAttributes def
+
+instance ParseAttributes SMTAttributes where
+    attributesParser = do
+        hook <- attributesParser
+        pure SMTAttributes {..} 
+
+-- | Returns `Just True` if the sentence is satisfied in all models,
+-- `Just False` if it has a counterexample,
+-- Nothing if timeout/undecidable.
+provePattern
+    :: Given (MetadataTools Object SMTAttributes)
+    => CommonPurePattern Object
+    -> Maybe Bool
+provePattern p =
+    case res of
+      Unsatisfiable _   -> Just True
+      Satisfiable   _ _ -> Just False
+      _ -> Nothing
+      where ThmResult res = unsafePerformIO $ provePatternIO p
+
+-- | Prints the result of SMT solving a pattern to the command line.
+provePatternIO
+    :: Given (MetadataTools Object SMTAttributes)
+    => CommonPurePattern Object
+    -> IO ThmResult
+provePatternIO = prove . patternToSMT
+
+patternToSMT
+    :: Given (MetadataTools Object SMTAttributes)
+    => CommonPurePattern Object
+    -> Predicate
+patternToSMT p = goTranslate
+  where
+    vars = Set.toList $ freeVars p
+    filterVars hookName =
+        filter (\v -> (getSortHook $ variableSort v) == Hook (Just hookName)) vars
+    boolVars = filterVars "BOOL.Bool"
+    intVars  = filterVars "INT.Int"
+    goTranslate = do
+        boolSMTVars <- sBools $ map (getId . variableName) boolVars
+        let boolTable = Map.fromList $ zip boolVars boolSMTVars
+        intSMTVars <- sIntegers $ map (getId . variableName) intVars
+        let intTable = Map.fromList $ zip intVars intSMTVars
+        go boolTable intTable
+    go boolTable intTable = goBoolean p
+      where
+        goUnaryOp op cont x1 =
+          do
+          tx1 <- cont x1
+          return (op tx1)
+        goBinaryOp op cont x1 x2 =
+          do
+          tx1 <- cont x1
+          tx2 <- cont x2
+          return (tx1 `op` tx2)
+        goBoolean (And_     _   x1 x2) = goBinaryOp (&&&) goBoolean x1 x2
+        goBoolean (Or_      _   x1 x2) = goBinaryOp (|||) goBoolean x1 x2
+        goBoolean (Implies_ _   x1 x2) = goBinaryOp (==>) goBoolean x1 x2
+        goBoolean (Iff_     _   x1 x2) = goBinaryOp (<=>) goBoolean x1 x2
+        goBoolean (Not_     _   x1)    = goUnaryOp (bnot) goBoolean x1
+        goBoolean (Equals_  s _ x1 x2) = c x1 x2
+          where
+            c = case getHookString $ getSortHook s of
+                "BOOL.Bool" -> goBinaryOp (.==) goInteger
```

Comment:
Aren't `INT.Int` and `BOOL.Bool` switched?

Reply:
Evidently do need more tests..